Change qtgraph and cbtf-argonavis to create proper module file library paths

jgalarowicz opened the pull request:
Fix bug where we only supported lib64 when we needed to look at lib on some platforms and lib64 on others.  Found when trying to build on Power systems.

adamjstewart commented on var/spack/repos/builtin/packages/qtgraph/package.py:
```diff
@@ -71,9 +72,13 @@ def setup_environment(self, spack_env, run_env):
+        if os.path.isdir(self.prefix.lib64):
+            lib_dir = self.prefix.lib64
+        else:
+            lib_dir = self.prefix.lib
```
@alalazo Do you know if there is a way to use `self.spec.libs` here? Obviously this dir won't exist at build time, but it will exist at module creation time, correct? We really should separate out `setup_environment` into two separate functions someday.

jgalarowicz replied:
@adamjstewart I'm trying the self.spec.libs suggestion now.   Thanks!   It looks like it is not being used in any packages, when I did a grep for it.

adamjstewart replied:
It's used in quite a few packages, but it is normally used to look at dependencies, not at the package itself.
```console
$ grep spec.*libs */package.py | wc -l
     232
```

alalazo replied:
If the implementation didn't change in the meanwhile, the key is the subscript operator:
```python
self.spec['qtgraph'].libs
```
should work as expected.

jgalarowicz replied:
@adamjstewart This is what I get:
```
+] /home/jeg/OSS_SPACK/spack/opt/spack/linux-fedora27-x86_64/gcc-7.2.1/qt-5.10.0-qh3dwgmgclmsrdx4al7vmjtqhl56yjb4
==> Installing qtgraph
==> Error: AttributeError: 'Spec' object has no attribute 'libs'

/home/jeg/OSS_SPACK/spack/var/spack/repos/builtin/packages/qtgraph/package.py:82, in setup_environment:
         79
         80        run_env.prepend_path(
         81            'LD_LIBRARY_PATH', join_path(
  >>     82                self.spec.libs, '{0}'.format(self.spec['qt'].version.up_to(3))))
         83
         84        run_env.prepend_path('CPATH', self.prefix.include.QtGraph)
```

alalazo replied:
@jgalarowicz See my comment above.

jgalarowicz replied:
@alalazo @adamjstewart Now get this:
```
==> qt is already installed in /home/jeg/OSS_SPACK/spack/opt/spack/linux-fedora27-x86_64/gcc-7.2.1/qt-5.10.0-qh3dwgmgclmsrdx4al7vmjtqhl56yjb4
==> Installing qtgraph
==> Error: RuntimeError: Unable to recursively locate qtgraph libraries in /home/jeg/OSS_SPACK/spack/opt/spack/linux-fedora27-x86_64/gcc-7.2.1/qtgraph-1.0.0.0-ddoww2gcrpfybd5rg2a3ewe2ct6t56gl

/home/jeg/OSS_SPACK/spack/var/spack/repos/builtin/packages/qtgraph/package.py:82, in setup_environment:
         79
         80        run_env.prepend_path(
         81            'LD_LIBRARY_PATH', join_path(
  >>     82                self.spec['qtgraph'].libs, '{0}'.format(self.spec['qt'].version.up_to(3))))
         83
         84        run_env.prepend_path('CPATH', self.prefix.include.QtGraph)


[jeg@localhost packages]$ 
```

jgalarowicz replied:
@alalazo @adamjstewart I guess qtgraph isn't installed yet.  That path doesn't exist yet.

adamjstewart replied:
Yeah, the problem is that `setup_environment` is run before build time (to set `spack_env`) and after installation (to set `run_env`). You may have to add a try-except statement. I had to do something like this for all of the intel packages.

jgalarowicz replied:
This works.  It is what you have in the intel packages I looked at.  Thanks for the pointers!
```    68	    def setup_environment(self, spack_env, run_env):
    69	        """Set up the compile and runtime environments for a package."""
    70	        spack_env.set('GRAPHVIZ_ROOT', self.spec['graphviz'].prefix)
    71	        spack_env.set('INSTALL_ROOT', self.prefix)
    72	
    73	        # NOTE: Spack runs setup_environment twice, once pre-build to set up
    74	        # the build environment, and once post-installation to determine
    75	        # the environment variables needed at run-time to add to the module
    76	        # file. The script we need to source is only present post-installation,
    77	        # so check for its existence before sourcing.
    78	        # TODO: At some point we should split setup_environment into
    79	        # setup_build_environment and setup_run_environment to get around
    80	        # this problem.
    81	
    82	        # The implementor has set up the library and include paths in
    83	        # a non-conventional way.  We reflect that here.
    84	        qtgraphlibs = join_path(
    85	            self.spec['qtgraph'].libs,
    86	            '{0}'.format(self.spec['qt'].version.up_to(3)))
    87	
    88	        if os.path.isfile(qtgraphlibs):
    89	            run_env.prepend_path(
    90	                'LD_LIBRARY_PATH', qtgraphlibs)
    91	            run_env.prepend_path('CPATH', self.prefix.include.QtGraph)
```

jgalarowicz replied:
@adamjstewart I haven't pushed the change yet.  I will wait on your advice.

jgalarowicz replied:
@adamjstewart I pushed the change, since I think it will be what you wanted.  If not, I can change the fix.

adamjstewart commented on var/spack/repos/builtin/packages/qtgraph/package.py:
```diff
@@ -69,11 +70,22 @@ def setup_environment(self, spack_env, run_env):
+        # NOTE: Spack runs setup_environment twice, once pre-build to set up
+        # the build environment, and once post-installation to determine
+        # the environment variables needed at run-time to add to the module
+        # file. The script we need to source is only present post-installation,
```
This needs to be reworded. You aren't sourcing any script.

adamjstewart commented on var/spack/repos/builtin/packages/qtgraph/package.py:
```diff
@@ -69,11 +70,22 @@ def setup_environment(self, spack_env, run_env):
-                self.prefix.lib64,
-                '{0}'.format(self.spec['qt'].version.up_to(3))))
+        qtgraphlibs = join_path(
+            self.spec['qtgraph'].libs,
```
Libs should return a list of library files, not the lib or lib64 directory. Take a look at the `LibraryList` docs.

adamjstewart replied:
I forget, does libs raise an error if it can't find anything or does it just return None/an empty list? You shouldn't need to check if the path exists, it can't return a non-existing path. Just check if it's empty.

adamjstewart commented on var/spack/repos/builtin/packages/qtgraph/package.py:
```diff
@@ -69,11 +70,22 @@ def setup_environment(self, spack_env, run_env):
+        if os.path.isfile(qtgraphlibs):
+            run_env.prepend_path(
+                'LD_LIBRARY_PATH', qtgraphlibs)
+            run_env.prepend_path('CPATH', self.prefix.include.QtGraph)
```
This line doesn't need to be in the if statement.

jgalarowicz commented:
@adamjstewart I reverted back to my original checked code and added similar code to cbtf-argonavis-gui, which also references the funky qtgraph library path in order to work properly.   I could not get the libs method to work after several frustrating hours.

jgalarowicz commented:
@adamjstewart Was the failure - a test failure (timeout) or do you think something in the commit caused this problem?   Thanks!

adamjstewart commented:
Looks like a random timeout. I restarted the build.

jgalarowicz commented:
@tgamblin  Hi Todd - I don't see a long commit message.  Can you point me to it?
```
in addition to @adamjstewart's requests can you please fix the commit message so that it's not so long? See https://chris.beams.io/posts/git-commit/ -- the details should be after the first line.
```

adamjstewart commented:
@jgalarowicz He's referring to your first (and probably third) commit messages. Both are longer than 50 characters. This causes the PR title to wrap onto multiple lines and the [commit messages](https://github.com/spack/spack/pull/8797/commits) to end up with `...` at the end.

See https://help.github.com/articles/changing-a-commit-message/ for instructions on how to change a commit message.

jgalarowicz commented:
@adamjstewart How do you change a commit that is already "committed"?    When I try the pencil looking icon, I don't see a way to change the commit message.   Thanks and sorry for not following the git check in protocol - I always think of it after I do the commit, unfortunately.

adamjstewart commented:
@jgalarowicz See the "Amending older of multiple commit messages" section of https://help.github.com/articles/changing-a-commit-message/. You'll have to do it from the command-line.

jgalarowicz commented:
@adamjstewart @tgamblin Thanks for the pointer - I hope it worked.  I followed the reword scenario.

jgalarowicz commented:
@adamjstewart @tgamblin Did my last changes satisfy your requests?   Thanks
